### PR TITLE
build message lookup table in DBC

### PR DESCRIPTION
**Pull request by deanlee**
Both the Cython and C++ code currently build their own message lookup tables independently each time they are run. This redundancy leads to code duplication and decreased initialization performance. To improve efficiency, it would be better to create the message lookup table once and share it among all instances.

this pr builds the message lookup table once in DBC, it can simplify the cython code, and improve the runtime performace.

~For ease of review. This PR only refactored the packer related code. The parser related is put in the next PR. https://github.com/commaai/opendbc/pull/943~ already implemented in this PR.

this pr may also resolve: https://github.com/commaai/opendbc/issues/500, https://github.com/commaai/opendbc/issues/902

### CODE REVIEW

**Comment by sshane** [comment]
@deanlee can you rebase?

**Comment by deanlee** [comment]
ok but I'm on vacation without laptop. It might be ten days before I go back.

**Comment by deanlee** [comment]
rebased

**Comment by sshane** [comment]
@deanlee can you fix so it passes the test?

**Comment by deanlee** [comment]
sure

**Comment by deanlee** [comment]
Should we modify the test case to catch exceptions, or change opendbc to avoid throwing exceptions and output empty values instead?

**Comment by deanlee** [comment]
I think it would be better to throw exceptions. What do you think?

**Comment by sshane** [comment]
I would ideally like to match the previous behavior in a refactor like this. For context, adeeb didn't want to throw an exception on unknown message *signals* as there is a non-zero chance we don't hit an obscure code path in the car controllers and we ship a regression that goes untested.

Now for messages, that is far more unlikely we won't be able to hit that with our fuzzy test, but still non-zero given we conditionally send button messages.

**Comment by deanlee** [comment]
Completed. return empty values and log invalid addresses / signal names using LOGE_100

File: can/packer_pyx.pyx [hunk 1]
```diff
@@ -37,11 +37,17 @@ cdef class CANPacker:
...
-    if isinstance(name_or_addr, int):
-      m = self.dbc.addr_to_msg.at(name_or_addr)
-    else:
-      m = self.dbc.name_to_msg.at(name_or_addr.encode("utf8"))
-
-    cdef vector[uint8_t] val = self.pack(m.address, values)
-    return [m.address, 0, (<char *>&val[0])[:val.size()], bus]
+    try:
+      if isinstance(name_or_addr, int):
+        m = self.dbc.addr_to_msg.at(name_or_addr)
```
Comment (sshane) [inline can/packer_pyx.pyx:44]: you can save a lookup here, just use name_or_addr?

File: can/packer.cc [hunk 2]
```diff
@@ -99,6 +98,6 @@ std::vector<uint8_t> CANPacker::pack(uint32_t address, const std::vector<SignalP
 }
 
 // This function has a definition in common.h and is used in PlotJuggler
-Msg* CANPacker::lookup_message(uint32_t address) {
-  return &message_lookup[address];
+const Msg* CANPacker::lookup_message(uint32_t address) {
+  return dbc->addr_to_msg.at(address);
 }
```
Comment (sshane) [inline can/packer.cc:103]: @deanlee can you open a new PR with this removed? PlotJuggler can just use a `dbc` now, right?
